File: handson/dotnet/Controllers/OrderController.cs
using AwesomePizza.Database;
using AwesomePizza.Dto;
using AwesomePizza.Models;
using Microsoft.AspNetCore.Mvc;

namespace AwesomePizza.Controllers;

[ApiController]
[Route("api/orders")]
public class OrderController : ControllerBase
{
    private readonly InMemoryDatabase _database;

    public OrderController(InMemoryDatabase database)
    {
        _database = database;
    }

    [HttpGet("{id}")]
    public IActionResult GetOrder(string id)
    {
        try
        {
            var order = _database.FindOrderById(id);
            if (order is null)
            {
                return NotFoundResponse($"Order with ID '{id}' not found");
            }

            return Ok(new { success = true, data = order, message = "Order retrieved successfully" });
        }
        catch
        {
            return StatusCode(500, new { success = false, error = "Internal server error", message = "Failed to retrieve order" });
        }
    }

    [HttpPost]
    public IActionResult CreateOrder([FromBody] CreateOrderRequest request)
    {
        try
        {
            var sender = request.Sender;
            if (string.IsNullOrWhiteSpace(sender))
            {
                return BadRequestResponse("Order sender is required and must be a non-empty string");
            }

            var contents = request.Contents;
            if (contents is null || contents.Count == 0)
            {
                return BadRequestResponse("Order contents are required and must be a non-empty array");
            }

            var itemError = ValidateItems(contents);
            if (itemError is not null)
            {
                return BadRequestResponse(itemError);
            }

            var newOrder = _database.AddOrder(sender.Trim(), contents);
            return StatusCode(201, new { success = true, data = newOrder, message = "Order created successfully" });
        }
        catch
        {
            return StatusCode(500, new { success = false, error = "Internal server error", message = "Failed to create order" });
        }
    }

    [HttpPut("{id}")]
    public IActionResult UpdateOrder(string id, [FromBody] UpdateOrderRequest request)
    {
        try
        {
            var order = _database.FindOrderById(id);
            if (order is null)
            {
                return NotFoundResponse($"Order with ID '{id}' not found");
            }

            if (request.Sender is not null)
            {
                if (string.IsNullOrWhiteSpace(request.Sender))
                {
                    return BadRequestResponse("Order sender must be a non-empty string");
                }
                order.Sender = request.Sender.Trim();
            }

            if (request.Status is not null)
            {
                if (!Enum.TryParse<OrderStatus>(request.Status, false, out var status))
                {
                    var valid = string.Join(", ", Enum.GetNames<OrderStatus>());
                    return BadRequestResponse($"Status must be one of: {valid}");
                }
                order.Status = status;
            }

            if (request.Contents is not null)
            {
                if (request.Contents.Count == 0)
                {
                    return BadRequestResponse("Order contents must be a non-empty array");
                }

                var itemError = ValidateItems(request.Contents);
                if (itemError is not null)
                {
                    return BadRequestResponse(itemError);
                }
                order.Contents = request.Contents;
            }

            return Ok(new { success = true, data = order, message = "Order updated successfully" });
        }
        catch
        {
            return StatusCode(500, new { success = false, error = "Internal server error", message = "Failed to update order" });
        }
    }

    private static string? ValidateItems(List<OrderItem> items)
    {
        foreach (var item in items)
        {
            if (string.IsNullOrWhiteSpace(item.Name))
            {
                return "Each order item must have a valid name";
            }
            if (item.Quantity is null || item.Quantity <= 0)
            {
                return "Each order item must have a valid quantity (positive number)";
            }
        }
        return null;
    }

    private BadRequestObjectResult BadRequestResponse(string message) =>
        BadRequest(new { success = false, error = "Bad request", message });

    private NotFoundObjectResult NotFoundResponse(string message) =>
        NotFound(new { success = false, error = "Not found", message });
}

File: handson/java/src/main/java/com/awesomepizza/controller/OrderController.java
package com.awesomepizza.controller;

import com.awesomepizza.database.InMemoryDatabase;
import com.awesomepizza.dto.CreateOrderRequest;
import com.awesomepizza.dto.UpdateOrderRequest;
import com.awesomepizza.model.Order;
import com.awesomepizza.model.OrderItem;
import com.awesomepizza.model.OrderStatus;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/orders")
public class OrderController {

    private final InMemoryDatabase database;

    public OrderController(InMemoryDatabase database) {
        this.database = database;
    }

    @GetMapping("/{id}")
    public ResponseEntity<Map<String, Object>> getOrder(@PathVariable String id) {
        try {
            Optional<Order> order = database.findOrderById(id);
            if (order.isEmpty()) {
                return errorResponse(HttpStatus.NOT_FOUND, "Not found", "Order with ID '" + id + "' not found");
            }
            return successResponse(HttpStatus.OK, order.get(), "Order retrieved successfully");
        } catch (Exception e) {
            return errorResponse(HttpStatus.INTERNAL_SERVER_ERROR, "Internal server error", "Failed to retrieve order");
        }
    }

    @PostMapping
    public ResponseEntity<Map<String, Object>> createOrder(@RequestBody CreateOrderRequest request) {
        try {
            String sender = request.getSender();
            if (sender == null || sender.trim().isEmpty()) {
                return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", "Order sender is required and must be a non-empty string");
            }

            List<OrderItem> contents = request.getContents();
            if (contents == null || contents.isEmpty()) {
                return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", "Order contents are required and must be a non-empty array");
            }

            String itemError = validateItems(contents);
            if (itemError != null) {
                return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", itemError);
            }

            Order newOrder = database.addOrder(sender.trim(), contents);
            return successResponse(HttpStatus.CREATED, newOrder, "Order created successfully");
        } catch (Exception e) {
            return errorResponse(HttpStatus.INTERNAL_SERVER_ERROR, "Internal server error", "Failed to create order");
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<Map<String, Object>> updateOrder(@PathVariable String id, @RequestBody UpdateOrderRequest request) {
        try {
            Optional<Order> existing = database.findOrderById(id);
            if (existing.isEmpty()) {
                return errorResponse(HttpStatus.NOT_FOUND, "Not found", "Order with ID '" + id + "' not found");
            }

            Order order = existing.get();

            if (request.getSender() != null) {
                if (request.getSender().trim().isEmpty()) {
                    return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", "Order sender must be a non-empty string");
                }
                order.setSender(request.getSender().trim());
            }

            if (request.getStatus() != null) {
                OrderStatus status;
                try {
                    status = OrderStatus.valueOf(request.getStatus());
                } catch (IllegalArgumentException e) {
                    String valid = Arrays.stream(OrderStatus.values()).map(Enum::name).collect(Collectors.joining(", "));
                    return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", "Status must be one of: " + valid);
                }
                order.setStatus(status);
            }

            if (request.getContents() != null) {
                if (request.getContents().isEmpty()) {
                    return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", "Order contents must be a non-empty array");
                }
                String itemError = validateItems(request.getContents());
                if (itemError != null) {
                    return errorResponse(HttpStatus.BAD_REQUEST, "Bad request", itemError);
                }
                order.setContents(request.getContents());
            }

            return successResponse(HttpStatus.OK, order, "Order updated successfully");
        } catch (Exception e) {
            return errorResponse(HttpStatus.INTERNAL_SERVER_ERROR, "Internal server error", "Failed to update order");
        }
    }

    private String validateItems(List<OrderItem> items) {
        for (OrderItem item : items) {
            if (item.getName() == null || item.getName().trim().isEmpty()) {
                return "Each order item must have a valid name";
            }
            if (item.getQuantity() == null || item.getQuantity() <= 0) {
                return "Each order item must have a valid quantity (positive number)";
            }
        }
        return null;
    }

    private ResponseEntity<Map<String, Object>> successResponse(HttpStatus status, Object data, String message) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("success", true);
        body.put("data", data);
        body.put("message", message);
        return ResponseEntity.status(status).body(body);
    }

    private ResponseEntity<Map<String, Object>> errorResponse(HttpStatus status, String error, String message) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("success", false);
        body.put("error", error);
        body.put("message", message);
        return ResponseEntity.status(status).body(body);
    }
}
